packages/backend-platform/src/WebsocketService.ts

```diff
@@ -0,0 +1,1348 @@
+  | WebSocketServiceIsChannelSubscribedAction
+  | WebSocketServiceAddChannelCallbackAction
+  | WebSocketServiceRemoveChannelCallbackAction
+  | WebSocketServiceGetChannelCallbacksAction;
+
+export type WebSocketServiceAllowedActions = never;
+
+export type WebSocketServiceAllowedEvents = never;
+
+// Event types for WebSocket connection state changes
+export type WebSocketServiceConnectionStateChangedEvent = {
+  type: 'BackendWebSocketService:connectionStateChanged';
+  payload: [WebSocketConnectionInfo];
+};
+
+export type WebSocketServiceEvents =
+  WebSocketServiceConnectionStateChangedEvent;
+
+export type WebSocketServiceMessenger = RestrictedMessenger<
+  typeof SERVICE_NAME,
+  WebSocketServiceActions | WebSocketServiceAllowedActions,
+  WebSocketServiceEvents | WebSocketServiceAllowedEvents,
+  WebSocketServiceAllowedActions['type'],
+  WebSocketServiceAllowedEvents['type']
+>;
+
+/**
+ * WebSocket Service with automatic reconnection, session management and direct callback routing
+ *
+ * Real-Time Performance Optimizations:
+ * - Fast path message routing (zero allocations)
+ * - Production mode removes try-catch overhead
+ * - Optimized JSON parsing with fail-fast
+ * - Direct callback routing bypasses event emitters
+ * - Memory cleanup and resource management
+ *
+ * Mobile Integration:
+ * Mobile apps should handle lifecycle events (background/foreground) by:
+ * 1. Calling disconnect() when app goes to background
+ * 2. Calling connect() when app returns to foreground
+ * 3. Calling destroy() on app termination
+ */
+export class WebSocketService {
+  readonly #messenger: WebSocketServiceMessenger;
+
+  readonly #options: Required<WebSocketServiceOptions>;
+
+  #ws!: WebSocket;
+
+  #state: WebSocketState = WebSocketState.DISCONNECTED;
+
+  #reconnectAttempts = 0;
+
+  #reconnectTimer: NodeJS.Timeout | null = null;
+
+  // Track the current connection promise to handle concurrent connection attempts
+  #connectionPromise: Promise<void> | null = null;
+
+  readonly #pendingRequests = new Map<
+    string,
+    {
+      resolve: (value: unknown) => void;
+      reject: (error: Error) => void;
+      timeout: NodeJS.Timeout;
+    }
+  >();
+
+  #lastError: string | null = null;
+
+  #connectedAt: number | null = null;
+
+  // Simplified subscription storage (single flat map)
+  // Key: subscription ID string (e.g., 'sub_abc123def456')
+  // Value: InternalSubscription object with channels, callback and metadata
+  readonly #subscriptions = new Map<string, InternalSubscription>();
+
+  // Channel-based callback storage
+  // Key: channel name (serves as unique identifier)
+  // Value: ChannelCallback configuration
+  readonly #channelCallbacks = new Map<string, ChannelCallback>();
+
+  /**
+   * Creates a new WebSocket service instance
+   *
+   * @param options - Configuration options including messenger
+   */
+  constructor(
+    options: WebSocketServiceOptions & { messenger: WebSocketServiceMessenger },
```

Comment:
Nit: `messenger` is an option of this service, so instead of adding it here, should it be in `WebSocketServiceOptions`, and then you can use `Omit<WebSocketServiceOptions, 'messenger'>` as the type for the `#options` property?